pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/properties/EnumeratedProperty.java

```diff
@@ -4,97 +4,115 @@
+    implements EnumeratedPropertyDescriptor<E, E> {
 
-    public static final PropertyDescriptorFactory FACTORY = new BasicPropertyDescriptorFactory<EnumeratedProperty>(
-            Enumeration.class) {
+    /** Factory. */
+    public static final PropertyDescriptorFactory<? extends Enumeration> FACTORY // @formatter:off
+        = new SingleValuePropertyDescriptorFactory<Enumeration>(Enumeration.class) { // TODO:cf Enumeration? Object?
+
+            @Override
+            public EnumeratedProperty createWith(Map<PropertyDescriptorField, String> valuesById) {
+                return new EnumeratedProperty<>(nameIn(valuesById),
+                                                descriptionIn(valuesById),
+                                                labelsIn(valuesById),   // this is not implemented
+                                                choicesIn(valuesById),  // ditto
+                                                indexIn(valuesById),    // ditto
+                                                classIn(valuesById),
+                                                0f);
+            }
+        }; // @formatter:on
+
+
+    private final Map<String, E> choicesByLabel;
+    private final Map<E, String> labelsByChoice;
+    private final Class<E> valueType;
 
-        @Override
-        public EnumeratedProperty createWith(Map<String, String> valuesById) {
 
-            return new EnumeratedProperty(nameIn(valuesById), descriptionIn(valuesById), labelsIn(valuesById),
-                    choicesIn(valuesById), indexIn(valuesById), 0f);
-        }
-    };
-
-    /**
-     * Constructor for EnumeratedProperty.
-     *
-     * @param theName
-     *            String
-     * @param theDescription
-     *            String
-     * @param theLabels
-     *            String[]
-     * @param theChoices
-     *            E[]
-     * @param defaultIndex
-     *            int
-     * @param theUIOrder
-     *            float
-     * @throws IllegalArgumentException
-     */
     public EnumeratedProperty(String theName, String theDescription, String[] theLabels, E[] theChoices,
-            int defaultIndex, float theUIOrder) {
-        super(theName, theDescription, theLabels, theChoices, new int[] { defaultIndex }, theUIOrder, false);
+                              int defaultIndex, Class<E> valueType, float theUIOrder) {
+        this(theName, theDescription, CollectionUtil.mapFrom(theLabels, theChoices),
+             theChoices[defaultIndex], valueType, theUIOrder);
     }
 
-    /**
-     * @return Class
-     * @see net.sourceforge.pmd.PropertyDescriptor#type()
-     */
+
+    public EnumeratedProperty(String theName, String theDescription, Map<String, E> labelsToChoices,
+                              E defaultValue, Class<E> valueType, float theUIOrder) {
+        super(theName, theDescription, defaultValue, theUIOrder);
+
+        this.valueType = valueType;
```

Comment:
Yes in general, I wasn't considering there was a need for that kind of usage.  In the heterogeneous value type case, the class of one value may not even be assignable with another value.  One would have to choose amount the set of classes in the intersection set of all the value hierarchies, decidedly non-trivial.

I don't think the `E[]` constructor would have this same issue, the component type would be compatible.

Let's say you did one of these scenarios, what is the practical use of such a `type()` at this point, especially for an enumeration, as once is basically limited to the pre-defined choices?

That said, providing `Class<E>` explicitly would get the compiler involved to unsure a compatible value was provided for both constructors.